src/trace.c

```diff
@@ -2121,3 +2140,79 @@ static struct device_data *find_device (struct device_data **device_hash,
 
 	return dev;
 }
+
+static int
+maybe_insert_path (struct path_data **path_table, const char *pathname)
+{
+	assert (pathname != NULL);
+	assert (path_table != NULL);
+
+	/* Simple DJB2 hashing. */
+	size_t i;
+	unsigned long hash = 5381;
+	size_t length = strlen (pathname);
+
+	for (i = 0; i < length; i++)
+		hash = ((hash << 5) + hash) + (unsigned long)pathname[i];
+
+	int key = hash & HASH_MASK;
+
+	struct path_data *data;
+	for (data = path_table[key]; data; data = data->next) {
+		if (data->hash == hash &&
+		    data->length == length &&
+		    strcmp (pathname, data->path) == 0)
+			return 0;
+	}
+
+	/* Path not found. add a new one,
+	 * and insert it to the start of linked list.
+	 */
+	data = calloc (1, sizeof (struct path_data));
+	assert (data != NULL);
+
+	data->hash = hash;
+	data->length = length;
+
+	strncpy (data->path, pathname, PACK_PATH_MAX);
+	data->path[PACK_PATH_MAX] = '\0';
+
+	if (path_table[key])
```

Comment:
You are not saving anything with this compare. Control flow can actually be more expensive than an assignment to hot cache. Just assign `data->next = path_table[key];`. It makes the code easier to read as well.